### DISCUSSION TITLE
Understanding port handling / groups / individual position

### DISCUSSION THREAD

**Discussion by grafandreas** [clientIO/joint#2004] [Q&A]
Dear all, 

I am struggling to understand port handling in some of the example. The chatbot example has the following lines in app.shapes.ts

```
const outputPortPosition = (portsArgs: dia.Element.Port[], elBBox: dia.BBox): g.Point[] => {
    const step = OUT_PORT_WIDTH + PADDING_S;
    return portsArgs.map((port: dia.Element.Port, index: number) => new g.Point({
        x: PADDING_L + OUT_PORT_WIDTH / 2 + index * step,
        y: elBBox.height
    }));
};
```
However, in my derived code (patched the example a bit), it seems that I am not being passed a list of ports, but a function, because I defined a position: func for the port group:

```
out: {
                position: outputPortPosition,
}
```

and, no matter what I try, because it seems that ports.mjs always overwrites port.position with the group position in ports.mjs:_evaluatePort() and thus

ports.mjs in line 48

```
        var portsArgs = ports.map(function(port) {
            return port && port.position && port.position.args;
        });
```
will always return a list of the group position function.

Q: Did I overlook something. Is there a way to get the actual ports passed to the function? If no, is there a way to get the "real" element / list of ports from within the position function in order to find the port based on the index in the list (i.e. hack around the behavior)?

Any help appreciated,

Andreas

**Comment by kumilingus** [comment]
What are you trying to do? 😄 (sorry to keep asking the same question, but the context is important)

```js
const portLayout = (portsArgs) => {  console.log(portsArgs);  }
element.set({
    ports: {
        groups: {
            g1: {
                position: portLayout
            }
        },
        items: [{
            id: 'p1',
            group: 'g1',
            args: {
                port: 'p1'
            }
        }, {
            id: 'p2',
            group: 'g1',
            args: {
                port: 'p2'
            }
        }]
    }
});
// [{ port: 'p1', fn: portLayout }, { port: 'p2', fn: portLayout }]
```
